File: algorithms/strings/z-algorithm/java/ZFunction.java
public class ZFunction {

    public static int[] zFunction(int[] arr) {
        int n = arr.length;
        int[] z = new int[n];
        int l = 0, r = 0;
        for (int i = 1; i < n; i++) {
            if (i < r) {
                z[i] = Math.min(r - i, z[i - l]);
            }
            while (i + z[i] < n && arr[z[i]] == arr[i + z[i]]) {
                z[i]++;
            }
            if (i + z[i] > r) {
                l = i;
                r = i + z[i];
            }
        }
        return z;
    }
}

File: algorithms/strings/z-algorithm/typescript/zFunction.ts
export function zFunction(arr: number[]): number[] {
    const n = arr.length;
    const z = new Array(n).fill(0);
    let l = 0, r = 0;
    for (let i = 1; i < n; i++) {
        if (i < r) {
            z[i] = Math.min(r - i, z[i - l]);
        }
        while (i + z[i] < n && arr[z[i]] === arr[i + z[i]]) {
            z[i]++;
        }
        if (i + z[i] > r) {
            l = i;
            r = i + z[i];
        }
    }
    return z;
}
